src/lib.rs

```diff
@@ -163,32 +210,32 @@ use std::sync::atomic::{AtomicUsize, ATOMIC_USIZE_INIT, Ordering};
 #[macro_use]
 mod macros;
 
-// The setup here is a bit weird to make at_exit work.
+// The setup here is a bit weird to make shutdown_logger work.
 //
 // There are four different states that we care about: the logger's
 // uninitialized, the logger's initializing (set_logger's been called but
-// LOGGER hasn't actually been set yet), the logger's active, or the logger's
-// shutting down inside of at_exit.
+// LOGGER hasn't actually been set yet), the logger's active, or the logger is
+// shut down after calling shutdown_logger.
 //
-// The LOGGER static is normally a Box<Box<Log>> with some special possible
-// values as well. The uninitialized and initializing states are represented by
-// the values 0 and 1 respectively. The shutting down state is also represented
-// by 1. Any other value is a valid pointer to the logger.
+// The LOGGER static holds a pointer to the global logger. It is protected by
+// the STATE static which determines whether LOGGER has been initialized yet.
 //
-// The at_exit routine needs to make sure that no threads are actively logging
-// when it deallocates the logger. The number of actively logging threads is
-// tracked in the REFCOUNT static. The routine first sets LOGGER back to 1.
+// The shutdown_logger routine needs to make sure that no threads are actively
+// logging before it returns. The number of actively logging threads is tracked
+// in the REFCOUNT static. The routine first sets STATE back to INITIALIZING.
 // All logging calls past that point will immediately return without accessing
 // the logger. At that point, the at_exit routine just waits for the refcount
 // to reach 0 before deallocating the logger. Note that the refcount does not
-// necessarily monotonically decrease at this point, as new log calls still
-// increment and decrement it, but the interval in between is small enough that
-// the wait is really just for the active log calls to finish.
-static LOGGER: AtomicUsize = ATOMIC_USIZE_INIT;
+// necessarily monotonically decrease at this point, as new  log calls still
+// increment and decrement it, but the interval in between is  small enough
+// that the wait is really just for the active log calls to finish.
+static mut LOGGER: Option<*const Log> = None;
```

Comment:
The original code used a `Box<Box<Log>>` in an atomic. The outer box is necessary because the inner one is a fat pointer, which doesn't fit in an atomic. The new code just guards a non-atomic fat pointer using an atomic integer.